keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1782 @@
+support updates because of runtime limitations, yet.
+
+```diff
+ message UpdateContainerResourcesRequest {
+
+ ...
+     // resources to update or other options to use when updating the container.
+     map<string, string> annotations = 4;
++    // Configuration of class resources.
++    ContainerClassResources class_resources = 5;
+}
+```
+
+#### PodSandboxConfig
+
+The `PodSandboxConfig` will be supplemented with a new `class_resources` field
+that specifies the assignment of pod-level QoS-class resources. The intended
+use for this would be to be able to communicate pod-level QoS-class resource
+assignments at sandbox creation time (`RunPodSandboxRequest`).
+
+```diff
+ message PodSandboxConfig {
+@@ -45,5 +45,14 @@ message PodSandboxConfig {
+     LinuxPodSandboxConfig linux = 8;
+     // Optional configurations specific to Windows hosts.
+     WindowsPodSandboxConfig windows = 9;
++    // Configuration of QoS-class resources.
++    PodClassResources class_resources = 10;
+ }
+
++// PodClassResources specifies the configuration of QoS-class resources
++// resources of a pod.
++message PodClassResources {
++    // QoS-class resource assignment of the pod.
++    // Key is the resource type and values is the class name within the resource type.
++    map<string, string> class = 1;
++}
+```
+
+#### RuntimeStatus
+
+Extend the `RuntimeStatus` message with new `resources` field that is used to
+communicate the available QoS-class resources from the runtime to the client.
+
+This information can be used by the client (kubelet) to validate QoS-class
+resource assignments before starting a pod. In future steps kubelet will patch
+this information into node status.
+
+```diff
+ message RuntimeStatus {
+     // List of current observed runtime conditions.
+     repeated RuntimeCondition conditions = 1;
++    // Information about the discovered resources
++    ResourcesInfo resources = 2;
++}
+
++// ResourcesInfo contains information about the resources discovered by the
++// runtime.
++message ResourcesInfo {
++    // Pod-level QoS-class resources available.
++    repeated ClassResourceInfo pod_class_resources = 1;
++    // Container-level QoS-class resources available.
++    repeated ClassResourceInfo container_class_resources = 2;
++}
+
++// ClassResourceInfo contains information about one type of class resource.
++message ClassResourceInfo {
++    // Name of the QoS-class resources.
++    string Name = 1;
++    // Mutable is set to true if this resource sypports dynamic updates.
++    bool Mutable = 2;
++    // List of classes of this resource.
++    repeated ClassResourceClassInfo classes = 3;
++}
+
++// ClassResourceClassInfo contains information about single class of one
++// QoS-class resource type.
++message ClassResourceClassInfo {
++    string Name = 1;
+ }
+```
+
+#### Consts
+
+Also, define "known" QoS-class resource types to more easily align container
+runtime implementations:
+
+```diff
++const (
```

Comment:
> If a particular runtime wants to define it's own resource that other runtimes don't support, does it need to add the resource to the list here?

No, that was not the intention.

Or is this mainly a way to provide some portability across runtimes for some common scenarios?

Yes, that was the thinking. This could be also moved elsewhere. I don't know if there's any good place to share these kind of definitions between runtimes, though 🤔